src/main/java/com/google/api/codegen/metadatagen/py/PythonPackageCopier.java

```diff
@@ -0,0 +1,120 @@
+    ImmutableMap.Builder<String, Doc> docBuilder = new ImmutableMap.Builder<String, Doc>();
+    List<String> pythonNamespacePackages = new ArrayList<>();
+    Path inputPath;
+    String apiVersion;
+
+    /**
+     * Constructor.
+     *
+     * @param inputPath The path to the (unprocessed) gRPC source code.
+     * @param apiVersion The major version of the API.
+     */
+    public PythonPackageFileVisitor(Path inputPath, String apiVersion) {
+      this.inputPath = inputPath;
+      this.apiVersion = apiVersion;
+    }
+
+    @Override
+    public FileVisitResult visitFile(Path file, BasicFileAttributes attr) throws IOException {
+      String outFile = inputPath.relativize(file).toString();
+      docBuilder.put(
+          outFile, Doc.text(new String(Files.readAllBytes(file), StandardCharsets.UTF_8)));
+      return FileVisitResult.CONTINUE;
+    }
+
+    @Override
+    public FileVisitResult postVisitDirectory(Path dir, IOException exc) {
+      // Top-level dir doesn't need an __init__.py
+      if (dir.equals(inputPath)) {
+        return FileVisitResult.CONTINUE;
+      }
+
+      String outFile = inputPath.relativize(dir.resolve("__init__.py")).toString();
+      // Version directory gets an empty __init__.py
+      if (dir.getFileName().toString().equals(apiVersion)) {
+        docBuilder.put(outFile, Doc.text("\n"));
+
+        // All others get become namespace packages
+      } else {
+        docBuilder.put(
+            outFile, Doc.text("__import__('pkg_resources').declare_namespace(__name__)\n"));
+        pythonNamespacePackages.add(inputPath.relativize(dir).toString());
+      }
+      return FileVisitResult.CONTINUE;
+    }
+
+    public List<String> getNamespacePackages() {
+      return pythonNamespacePackages;
+    }
+
+    public ImmutableMap.Builder<String, Doc> getDocBuilder() {
+      return docBuilder;
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public Map<String, Doc> run(ToolOptions options, String copierMetadataKey) throws IOException {
+    // Copy files from dir into map, and fill in namespace result
+    // Run __init__ snippet in each dir that deserves it
+    PythonPackageFileVisitor visitor =
+        new PythonPackageFileVisitor(
+            Paths.get(options.get(PackageMetadataGenerator.INPUT_DIR)),
+            options.get(PackageMetadataGenerator.API_VERSION));
+
+    Files.walkFileTree(Paths.get(options.get(PackageMetadataGenerator.INPUT_DIR)), visitor);
+
+    List<String> pythonNamespacePackages = visitor.getNamespacePackages();
+    ImmutableMap.Builder<String, Doc> docBuilder = visitor.getDocBuilder();
+
+    StringBuilder namespacePackages = new StringBuilder("[");
+    for (String namespacePackage : pythonNamespacePackages) {
+      namespacePackages.append("'" + namespacePackage + "', ");
+    }
+    namespacePackages.append("]");
+
+    docBuilder.put(copierMetadataKey, Doc.text(namespacePackages.toString()));
```

Comment:
Why not just create a simple value type that contains `Map<String,Doc>` and the namespace packages? Heterogeneous key/value bags are extremely difficult to maintain because it's really hard to trace how things are injected.